File: internal/scheduler/scheduler.go
package scheduler

import (
	"errors"
	"fmt"
	"sort"
	"sync/atomic"

	"github.com/Georgecane/openhost/internal/fabric"
	"github.com/Georgecane/openhost/internal/node"
	"github.com/Georgecane/openhost/internal/resource"
)

var (
	ErrInsufficientResources = errors.New("insufficient resources in fabric")
	ErrNilFabric             = errors.New("fabric must not be nil")
)

type Scheduler interface {
	Plan(requirement resource.ResourceFragment) (node.LogicalNode, error)
}

// AggregatingScheduler builds logical nodes from multiple resource offers.
type AggregatingScheduler struct {
	fabric fabric.Fabric
	serial atomic.Uint64
}

func NewAggregatingScheduler(f fabric.Fabric) (*AggregatingScheduler, error) {
	if f == nil {
		return nil, ErrNilFabric
	}
	return &AggregatingScheduler{fabric: f}, nil
}

func (s *AggregatingScheduler) Plan(requirement resource.ResourceFragment) (node.LogicalNode, error) {
	if err := resource.ValidateRequirement(requirement); err != nil {
		return node.LogicalNode{}, err
	}

	offers := append([]fabric.ResourceOffer(nil), s.fabric.Offers()...)
	sort.Slice(offers, func(i, j int) bool {
		return offers[i].ParticipantID < offers[j].ParticipantID
	})

	var total resource.ResourceFragment
	allocations := make([]node.Allocation, 0, len(offers))

	for _, offer := range offers {
		if offer.ParticipantID == "" || offer.Resources.Empty() {
			continue
		}

		remaining := remainingResources(requirement, total)
		share := minFragment(offer.Resources, remaining)
		if share.Empty() {
			continue
		}

		allocations = append(allocations, node.Allocation{
			ParticipantID: offer.ParticipantID,
			Resources:     share,
		})
		total = total.Add(share)

		if total.Satisfies(requirement) {
			break
		}
	}

	if !total.Satisfies(requirement) {
		return node.LogicalNode{}, fmt.Errorf(
			"%w: requested=%+v available=%+v",
			ErrInsufficientResources,
			requirement,
			total,
		)
	}

	return node.LogicalNode{
		ID:          fmt.Sprintf("logical-%d", s.serial.Add(1)),
		Resources:   total,
		Allocations: allocations,
	}, nil
}

func remainingResources(requirement, allocated resource.ResourceFragment) resource.ResourceFragment {
	return resource.ResourceFragment{
		CPU: resource.CPUCapacity{
			Cores: maxFloat(requirement.CPU.Cores-allocated.CPU.Cores, 0),
		},
		Memory: resource.MemoryCapacity{
			Bytes: maxUint64(requirement.Memory.Bytes, allocated.Memory.Bytes) - allocated.Memory.Bytes,
		},
		Storage: resource.StorageCapacity{
			Bytes: maxUint64(requirement.Storage.Bytes, allocated.Storage.Bytes) - allocated.Storage.Bytes,
		},
		Network: resource.NetworkCapacity{
			BitsPerSecond: maxUint64(requirement.Network.BitsPerSecond, allocated.Network.BitsPerSecond) - allocated.Network.BitsPerSecond,
		},
		GPU: resource.GPUCapacity{
			Units: maxUint32(requirement.GPU.Units, allocated.GPU.Units) - allocated.GPU.Units,
		},
		Lifetime: requirement.Lifetime,
	}
}

func minFragment(available, required resource.ResourceFragment) resource.ResourceFragment {
	if required.Lifetime > 0 && available.Lifetime > 0 && available.Lifetime < required.Lifetime {
		return resource.ResourceFragment{}
	}

	lifetime := required.Lifetime
	if available.Lifetime > 0 && (lifetime == 0 || available.Lifetime < lifetime) {
		lifetime = available.Lifetime
	}

	return resource.ResourceFragment{
		CPU: resource.CPUCapacity{Cores: minFloat(available.CPU.Cores, required.CPU.Cores)},
		Memory: resource.MemoryCapacity{
			Bytes: minUint64(available.Memory.Bytes, required.Memory.Bytes),
		},
		Storage: resource.StorageCapacity{
			Bytes: minUint64(available.Storage.Bytes, required.Storage.Bytes),
		},
		Network: resource.NetworkCapacity{
			BitsPerSecond: minUint64(available.Network.BitsPerSecond, required.Network.BitsPerSecond),
		},
		GPU: resource.GPUCapacity{
			Units: minUint32(available.GPU.Units, required.GPU.Units),
		},
		Lifetime: lifetime,
	}
}

func minFloat(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

func maxFloat(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}

func minUint64(a, b uint64) uint64 {
	if a < b {
		return a
	}
	return b
}

func maxUint64(a, b uint64) uint64 {
	if a > b {
		return a
	}
	return b
}

func minUint32(a, b uint32) uint32 {
	if a < b {
		return a
	}
	return b
}

func maxUint32(a, b uint32) uint32 {
	if a > b {
		return a
	}
	return b
}

File: src/scheduler.rs
use crate::fabric::{Fabric, ResourceOffer};
use crate::identity::{Identity, Kind};
use crate::node::LogicalNode;
use crate::resource::{Allocation, CompositeResource, CpuCapacity, GpuCapacity, MemoryCapacity, NetworkCapacity, ResourceFragment, StorageCapacity};
use std::sync::Arc;
use std::time::Duration;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SchedulerError { InsufficientResources, NilFabric }

impl std::fmt::Display for SchedulerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result { write!(f, "{self:?}") }
}
impl std::error::Error for SchedulerError {}

pub trait Scheduler: Send + Sync {
    fn plan(&self, requirement: ResourceFragment) -> Result<LogicalNode, SchedulerError>;
}

pub struct AggregatingScheduler {
    fabric: Arc<dyn Fabric>,
}

impl AggregatingScheduler {
    pub fn new(fabric: Arc<dyn Fabric>) -> Self { Self { fabric } }

    fn remaining(requirement: ResourceFragment, allocated: ResourceFragment) -> ResourceFragment {
        ResourceFragment {
            cpu: CpuCapacity { cores: (requirement.cpu.cores - allocated.cpu.cores).max(0.0) },
            memory: MemoryCapacity { bytes: requirement.memory.bytes.saturating_sub(allocated.memory.bytes) },
            storage: StorageCapacity { bytes: requirement.storage.bytes.saturating_sub(allocated.storage.bytes) },
            network: NetworkCapacity { bits_per_second: requirement.network.bits_per_second.saturating_sub(allocated.network.bits_per_second) },
            gpu: GpuCapacity { units: requirement.gpu.units.saturating_sub(allocated.gpu.units) },
            lifetime: requirement.lifetime,
        }
    }

    fn min_fragment(available: ResourceFragment, required: ResourceFragment) -> ResourceFragment {
        if let (Some(a), Some(r)) = (available.lifetime, required.lifetime) {
            if a < r { return ResourceFragment::default(); }
        }
        let lifetime = match (available.lifetime, required.lifetime) {
            (Some(a), Some(r)) => Some(a.min(r)),
            (Some(a), None) => Some(a),
            (None, Some(r)) => Some(r),
            (None, None) => None,
        };
        ResourceFragment {
            cpu: CpuCapacity { cores: available.cpu.cores.min(required.cpu.cores) },
            memory: MemoryCapacity { bytes: available.memory.bytes.min(required.memory.bytes) },
            storage: StorageCapacity { bytes: available.storage.bytes.min(required.storage.bytes) },
            network: NetworkCapacity { bits_per_second: available.network.bits_per_second.min(required.network.bits_per_second) },
            gpu: GpuCapacity { units: available.gpu.units.min(required.gpu.units) },
            lifetime,
        }
    }
}

impl Scheduler for AggregatingScheduler {
    fn plan(&self, requirement: ResourceFragment) -> Result<LogicalNode, SchedulerError> {
        requirement.validate_requirement().map_err(|_| SchedulerError::InsufficientResources)?;
        let mut offers: Vec<ResourceOffer> = self.fabric.offers();
        offers.sort_by(|a, b| a.participant_id.cmp(&b.participant_id));

        let mut allocations = Vec::new();
        let mut total = ResourceFragment::default();

        for offer in offers {
            if offer.participant_id.is_empty() || offer.resources.is_empty() { continue; }
            let share = Self::min_fragment(offer.resources, Self::remaining(requirement, total));
            if share.is_empty() { continue; }
            allocations.push(Allocation { participant_id: offer.participant_id, resources: share });
            total = total.add(share);
            if total.satisfies(&requirement) { break; }
        }

        if !total.satisfies(&requirement) { return Err(SchedulerError::InsufficientResources); }
        let composed = CompositeResource::compose(allocations).map_err(|_| SchedulerError::InsufficientResources)?;
        if !composed.satisfies(&requirement) { return Err(SchedulerError::InsufficientResources); }

        let id = Identity::new(Kind::LogicalNode);
        Ok(LogicalNode { id: id.id, resources: composed, runtime: Default::default() })
    }
}
